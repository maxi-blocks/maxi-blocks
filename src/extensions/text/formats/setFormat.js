/**
 * WordPress dependencies
 */
import { isCollapsed } from '@wordpress/rich-text';
import { dispatch } from '@wordpress/data';

/**
 * Internal dependencies
 */
import getFormattedString from './getFormattedString';
import getHasCustomFormat from './getHasCustomFormat';
import setFormatWithClass from './setFormatWithClass';
import flatFormatsWithClass from './flatFormatsWithClass';

/**
 * External dependencies
 */
import { isNil, isEmpty } from 'lodash';

/**
 *
 * @param {Object}  [$0]             Optional named arguments.
 * @param {Object}  [$0.formatValue] RichText format value
 * @param {Object}  [$0.typography]  MaxiBlocks typography
 * @param {boolean} [$0.isList]      Text Maxi block has list mode active
 * @param {Object}  [$0.value]       Requested values to implement
 *                                   on typography object
 * @param {string}  [$0.breakpoint]  Device type breakpoint
 * @param {boolean} [$0.isHover]     Is the requested typography under hover state
 * @returns {Object} Formatted object
 */
const setFormat = ({
	formatValue,
	typography,
	isList,
	value,
	breakpoint = 'general',
	isHover = false,
	textLevel,
	returnFormatValue = false,
	disableCustomFormats = false,
	styleCardPrefix = '',
}) => {
	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
	const { __unstableMarkLastChangeAsPersistent: markLastChangeAsPersistent } =
		dispatch('core/block-editor');

	if (disableCustomFormats) {
		const newTypography = { ...typography };

		Object.entries(value).forEach(([key, val]) => {
			newTypography[`${key}-${breakpoint}${isHover ? '-hover' : ''}`] =
				val;
		});
		// Ensures the format changes are saved as undo entity on historical records
		markLastChangeAsPersistent();

		return newTypography;
	}

	if (
		isNil(formatValue.start) ||
		isNil(formatValue.end) ||
		isCollapsed(formatValue)
	) {
		// eslint-disable-next-line @wordpress/no-global-get-selection, @wordpress/no-unguarded-get-range-at
		const { startOffset, endOffset } = window.getSelection().getRangeAt(0);

		formatValue.start = startOffset;
		formatValue.end = endOffset;
	}

	const formatsLength = !isEmpty(formatValue.formats)
		? formatValue.formats.length
		: 0;

	if (
		!formatValue ||
		formatValue.start === formatValue.end ||
		(formatValue.start === 0 && formatValue.end === formatsLength)
	) {
		const newTypography = { ...typography };
		const newFormatValue = {
			...formatValue,
			start: 0,
			end: formatsLength,
		};

		Object.entries(value).forEach(([key, val]) => {
			newTypography[`${key}-${breakpoint}${isHover ? '-hover' : ''}`] =
				val;
		});

		const hasCustomFormat = getHasCustomFormat(newFormatValue, isHover);

		if (hasCustomFormat) {
			const content = getFormattedString({
				formatValue: newFormatValue,
				isList,
			});

			const {
				typography: cleanedTypography,
				content: cleanedContent,
				formatValue: cleanedFormatValue,
			} = flatFormatsWithClass({
				formatValue: newFormatValue,
				typography: newTypography,
				content,
				isList,
				value,
				breakpoint,
				textLevel,
				isHover,
				styleCardPrefix,
				returnFormatValue,
			});

			return {
				...cleanedTypography,
				content: cleanedContent,
				...(returnFormatValue && {
					formatValue: {
						...cleanedFormatValue,
						start: formatValue.start,
						end: formatValue.end,
					},
				}),
			};
		}

		// Ensures the format changes are saved as undo entity on historical records
		markLastChangeAsPersistent();

		const newContent = newFormatValue.text
			? getFormattedString({
					formatValue: newFormatValue,
					isList,
			  })
			: false;

		return {
			...newTypography,
			...(newContent && { content: newContent }),
			...(returnFormatValue && {
				formatValue: {
					...newFormatValue,
					start: formatValue.start,
					end: formatValue.end,
				},
			}),
		};
	}

	// Ensures the format changes are saved as undo entity on historical records
	markLastChangeAsPersistent();

	return setFormatWithClass({
		formatValue,
		isList,
		typography,
		value,
		breakpoint,
		isHover,
		textLevel,
		returnFormatValue,
		styleCardPrefix,
	});
};

export default setFormat;
