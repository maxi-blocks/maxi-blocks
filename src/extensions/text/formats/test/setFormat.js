import setFormat from '@extensions/text/formats/setFormat';

jest.mock('@wordpress/data', () => ({
	dispatch: jest.fn(() => ({
		__unstableMarkLastChangeAsPersistent: jest.fn(),
	})),
	select: jest.fn(() => ({
		receiveMaxiSelectedStyleCard: jest.fn(() => ({ value: {} })),
	})),
}));
jest.mock('@extensions/styles', () => ({
	getBlockStyle: jest.fn(() => 'light'),
}));
jest.mock('@extensions/style-cards', () => ({
	updateSCOnEditor: jest.fn(),
	getTypographyFromSC: jest.fn(() => ({})),
}));

const colorFormatObject = {
	formatValue: {
		formats: Array(12),
		replacements: Array(12),
		text: 'Hello world!',
		start: 6,
		end: 11,
		activeFormats: [],
	},
	typography: {
		'palette-status-general': true,
		'palette-sc-status-general': false,
		'palette-color-general': 3,
	},
	value: {
		color: null,
		'palette-color': 4,
		'palette-status': true,
		'palette-sc-status': false,
		'palette-opacity': 1,
	},
	breakpoint: 'general',
	textLevel: 'p',
	returnFormatValue: true,
};

describe('setFormat', () => {
	let windowSpy;

	beforeEach(() => {
		windowSpy = jest.spyOn(window, 'window', 'get');
	});

	afterEach(() => {
		windowSpy.mockRestore();
	});

	it('Should correctly set the format when changing the color of part of the text', () => {
		const result = setFormat(colorFormatObject);

		expect(result).toMatchSnapshot();
	});

	it('Should just set and return the attributes when disableCustomFormats is true', () => {
		const result = setFormat({
			...colorFormatObject,
			value: {
				color: '#000',
			},
			disableCustomFormats: true,
		});

		expect(result).toEqual({
			'color-general': '#000',
		});
	});

	it('Should work when the selection is collapsed', () => {
		windowSpy.mockImplementation(() => ({
			getSelection: () => ({
				getRangeAt: () => ({
					startOffset: 1,
					endOffset: 1,
				}),
			}),
		}));

		const result = setFormat({
			...colorFormatObject,
			formatValue: {
				...colorFormatObject.formatValue,
				start: 1,
				end: 1,
			},
		});

		expect(result).toMatchSnapshot();
	});

	it('Should work when the selection is collapsed on existing format', () => {
		windowSpy.mockImplementation(() => ({
			getSelection: () => ({
				getRangeAt: () => ({
					startOffset: 7,
					endOffset: 7,
				}),
			}),
		}));

		const result = setFormat({
			...colorFormatObject,
			formatValue: {
				...colorFormatObject.formatValue,
				activeFormats: [
					{
						type: 'maxi-blocks/text-custom',
						attributes: {
							className: 'maxi-text-block__custom-format--0',
						},
					},
				],
				formats: Array(12).fill(
					[
						{
							type: 'maxi-blocks/text-custom',
							attributes: {
								className: 'maxi-text-block__custom-format--0',
							},
						},
					],
					6,
					11
				),
				start: 7,
				end: 7,
			},
			typography: {
				...colorFormatObject.typography,
				'custom-formats': {
					'maxi-text-block__custom-format--0': {
						'palette-color-general': 4,
						'palette-opacity-general': 1,
					},
				},
			},
			value: {
				...colorFormatObject.value,
				'palette-color': 5,
			},
		});

		expect(result).toMatchSnapshot();
	});
});
