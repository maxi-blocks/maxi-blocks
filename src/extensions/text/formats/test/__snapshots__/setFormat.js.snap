// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setFormat Should correctly set the format when changing the color of part of the text 1`] = `
{
  "content": "Hello <maxi-blocks/text-custom className="maxi-text-block__custom-format--0">world</maxi-blocks/text-custom>!",
  "custom-formats": {
    "maxi-text-block__custom-format--0": {
      "palette-color-general": 4,
      "palette-opacity-general": 1,
    },
  },
  "formatValue": {
    "activeFormats": [
      {
        "attributes": {
          "className": "maxi-text-block__custom-format--0",
        },
        "type": "maxi-blocks/text-custom",
      },
    ],
    "end": 11,
    "formats": [
      ,
      ,
      ,
      ,
      ,
      ,
      [
        {
          "attributes": {
            "className": "maxi-text-block__custom-format--0",
          },
          "type": "maxi-blocks/text-custom",
        },
      ],
      [
        {
          "attributes": {
            "className": "maxi-text-block__custom-format--0",
          },
          "type": "maxi-blocks/text-custom",
        },
      ],
      [
        {
          "attributes": {
            "className": "maxi-text-block__custom-format--0",
          },
          "type": "maxi-blocks/text-custom",
        },
      ],
      [
        {
          "attributes": {
            "className": "maxi-text-block__custom-format--0",
          },
          "type": "maxi-blocks/text-custom",
        },
      ],
      [
        {
          "attributes": {
            "className": "maxi-text-block__custom-format--0",
          },
          "type": "maxi-blocks/text-custom",
        },
      ],
      ,
    ],
    "replacements": [
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
    ],
    "start": 6,
    "text": "Hello world!",
  },
}
`;

exports[`setFormat Should work when the selection is collapsed 1`] = `
{
  "color-general": null,
  "content": "Hello world!",
  "formatValue": {
    "activeFormats": [],
    "end": 1,
    "formats": [
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
    ],
    "replacements": [
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
    ],
    "start": 1,
    "text": "Hello world!",
  },
  "palette-color-general": 4,
  "palette-opacity-general": 1,
  "palette-sc-status-general": false,
  "palette-status-general": true,
}
`;

exports[`setFormat Should work when the selection is collapsed on existing format 1`] = `
{
  "color-general": null,
  "content": "Hello world!",
  "custom-formats": {},
  "formatValue": {
    "activeFormats": [],
    "end": 7,
    "formats": [
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
    ],
    "replacements": [
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
      ,
    ],
    "start": 7,
    "text": "Hello world!",
  },
  "palette-color-general": 5,
  "palette-opacity-general": 1,
  "palette-sc-status-general": false,
  "palette-status-general": true,
}
`;
